File: 0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int i=0;
        int j=0;
        Set<Character> set = new HashSet<>();
        int maxLength = 0;

        while(j < n) {
            while(set.contains(s.charAt(j))) {
                set.remove(s.charAt(i));
                i++;//6
            }
            char ch = s.charAt(j);
            set.add(ch);//b
            //{}
            maxLength = Math.max(maxLength, j-i+1);//3
            j++;//7
        }
        return maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/solution.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var n = s.count
        var i=0
        var j=0
        var exists = Set<Character>()
        var chars = Array(s)
        var maxLength = 0

        while j < n {
            // invalid
            while exists.contains(chars[j]) {
                exists.remove(chars[i])
                i += 1
            }
            exists.insert(chars[j])//ab
            maxLength = max(maxLength, j-i+1)//1
            j += 1
        }
        return maxLength
    }
}
